File: features/step_definitions/patterns.rb
require_relative "../../lib/rayz/stripe_pattern"
require_relative "../../lib/rayz/gradient_pattern"
require_relative "../../lib/rayz/ring_pattern"
require_relative "../../lib/rayz/checkers_pattern"
require_relative "../../lib/rayz/test_pattern"

# Pattern creation steps
Given("pattern ← stripe_pattern\\({word}, {word})") do |color1, color2|
  @pattern = Rayz::StripePattern.new(instance_variable_get("@#{color1}"), instance_variable_get("@#{color2}"))
end

Given("pattern ← gradient_pattern\\({word}, {word})") do |color1, color2|
  @pattern = Rayz::GradientPattern.new(instance_variable_get("@#{color1}"), instance_variable_get("@#{color2}"))
end

Given("pattern ← ring_pattern\\({word}, {word})") do |color1, color2|
  @pattern = Rayz::RingPattern.new(instance_variable_get("@#{color1}"), instance_variable_get("@#{color2}"))
end

Given("pattern ← checkers_pattern\\({word}, {word})") do |color1, color2|
  @pattern = Rayz::CheckersPattern.new(instance_variable_get("@#{color1}"), instance_variable_get("@#{color2}"))
end

Given("pattern ← test_pattern\\()") do
  @pattern = Rayz::TestPattern.new
end

# Pattern property checks
Then("pattern.a = {word}") do |color|
  assert_equal(instance_variable_get("@#{color}"), @pattern.a)
end

Then("pattern.b = {word}") do |color|
  assert_equal(instance_variable_get("@#{color}"), @pattern.b)
end

# Pattern color evaluation
Then("stripe_at\\(pattern, point\\({float}, {float}, {float})) = {word}") do |x, y, z, color|
  point = Rayz::Point.new(x: x, y: y, z: z)
  result = @pattern.pattern_at(point)
  assert_equal(instance_variable_get("@#{color}"), result)
end

Then("pattern_at\\(pattern, point\\({float}, {float}, {float})) = {word}") do |x, y, z, color|
  point = Rayz::Point.new(x: x, y: y, z: z)
  result = @pattern.pattern_at(point)
  assert_equal(instance_variable_get("@#{color}"), result)
end

Then("pattern_at\\(pattern, point\\({float}, {float}, {float})) = color\\({float}, {float}, {float})") do |x, y, z, r, g, b|
  point = Rayz::Point.new(x: x, y: y, z: z)
  result = @pattern.pattern_at(point)
  expected = Rayz::Color.new(red: r, green: g, blue: b)
  assert_equal(expected, result)
end

# Pattern transformation
Then("pattern.transform = identity_matrix") do
  assert_equal(Matrix.identity(4), @pattern.transform)
end

When("set_pattern_transform\\(pattern, {transform})") do |transform_name|
  @pattern.transform = instance_variable_get("@#{transform_name}")
end

Then("pattern.transform = {word}") do |transform_name|
  assert_equal(instance_variable_get("@#{transform_name}"), @pattern.transform)
end

# Pattern with object transformation
When("c ← stripe_at_object\\(pattern, object, point\\({float}, {float}, {float}))") do |x, y, z|
  point = Rayz::Point.new(x: x, y: y, z: z)
  @c = @pattern.pattern_at_shape(@object, point)
end

When("c ← pattern_at_shape\\(pattern, shape, point\\({float}, {float}, {float}))") do |x, y, z|
  point = Rayz::Point.new(x: x, y: y, z: z)
  @c = @pattern.pattern_at_shape(@shape, point)
end

File: python/features/steps/patterns.py
from behave import given, then, use_step_matcher, when

from rayz.color import Color
from rayz.math_parser import parse_math
from rayz.matrix import Matrix
from rayz.pattern import (
    checkers_pattern,
    gradient_pattern,
    ring_pattern,
    stripe_pattern,
    test_pattern,
)
from rayz.transformations import scaling, translation
from rayz.tuple import Point

use_step_matcher("re")

_V = r"([A-Za-z][A-Za-z0-9_]*)"
_A = r"([^\s,)]+)"

IDENTITY_MATRIX = Matrix.identity(4)

_TRANSFORM_MAP = {"scaling": scaling, "translation": translation}


def _eval_transform(expr: str):
    import re

    m = re.fullmatch(r"(scaling|translation)\((.+)\)", expr.strip())
    if m:
        args = [parse_math(a.strip()) for a in m.group(2).split(",")]
        return _TRANSFORM_MAP[m.group(1)](*args)
    raise ValueError(f"Unknown transform: {expr!r}")


# ---------------------------------------------------------------------------
# Given
# ---------------------------------------------------------------------------


@given(rf"{_V} ← stripe_pattern\({_V},\s*{_V}\)")
def step_given_stripe_pattern(context, var, a_var, b_var):
    setattr(context, var, stripe_pattern(getattr(context, a_var), getattr(context, b_var)))


@given(rf"{_V} ← gradient_pattern\({_V},\s*{_V}\)")
def step_given_gradient_pattern(context, var, a_var, b_var):
    setattr(context, var, gradient_pattern(getattr(context, a_var), getattr(context, b_var)))


@given(rf"{_V} ← ring_pattern\({_V},\s*{_V}\)")
def step_given_ring_pattern(context, var, a_var, b_var):
    setattr(context, var, ring_pattern(getattr(context, a_var), getattr(context, b_var)))


@given(rf"{_V} ← checkers_pattern\({_V},\s*{_V}\)")
def step_given_checkers_pattern(context, var, a_var, b_var):
    setattr(context, var, checkers_pattern(getattr(context, a_var), getattr(context, b_var)))


@given(rf"{_V} ← test_pattern\(\)")
def step_given_test_pattern(context, var):
    setattr(context, var, test_pattern())


@given(rf"set_pattern_transform\({_V},\s*(.+)\)")
def step_given_set_pattern_transform(context, var, expr):
    getattr(context, var).set_transform(_eval_transform(expr))


# ---------------------------------------------------------------------------
# When
# ---------------------------------------------------------------------------


@when(rf"set_pattern_transform\({_V},\s*(.+)\)")
def step_when_set_pattern_transform(context, var, expr):
    getattr(context, var).set_transform(_eval_transform(expr))


@when(rf"c ← stripe_at_object\({_V},\s*{_V},\s*point\({_A},\s*{_A},\s*{_A}\)\)")
def step_when_stripe_at_object(context, pat_var, obj_var, x, y, z):
    pat = getattr(context, pat_var)
    obj = getattr(context, obj_var)
    context.c = pat.pattern_at_shape(obj, Point(parse_math(x), parse_math(y), parse_math(z)))


@when(rf"c ← pattern_at_shape\({_V},\s*{_V},\s*point\({_A},\s*{_A},\s*{_A}\)\)")
def step_when_pattern_at_shape(context, pat_var, shape_var, x, y, z):
    pat = getattr(context, pat_var)
    shape = getattr(context, shape_var)
    context.c = pat.pattern_at_shape(shape, Point(parse_math(x), parse_math(y), parse_math(z)))


# ---------------------------------------------------------------------------
# Then
# ---------------------------------------------------------------------------


@then(rf"{_V}\.a = {_V}")
def step_then_pattern_a(context, pat_var, color_var):
    assert getattr(context, pat_var).a == getattr(context, color_var)


@then(rf"{_V}\.b = {_V}")
def step_then_pattern_b(context, pat_var, color_var):
    assert getattr(context, pat_var).b == getattr(context, color_var)


@then(rf"stripe_at\({_V},\s*point\({_A},\s*{_A},\s*{_A}\)\) = {_V}")
def step_then_stripe_at_eq_var(context, pat_var, x, y, z, color_var):
    pat = getattr(context, pat_var)
    result = pat.pattern_at(Point(parse_math(x), parse_math(y), parse_math(z)))
    assert result == getattr(context, color_var)


@then(rf"pattern_at\({_V},\s*point\({_A},\s*{_A},\s*{_A}\)\) = {_V}")
def step_then_pattern_at_eq_var(context, pat_var, x, y, z, color_var):
    pat = getattr(context, pat_var)
    result = pat.pattern_at(Point(parse_math(x), parse_math(y), parse_math(z)))
    assert result == getattr(context, color_var)


@then(rf"pattern_at\({_V},\s*point\({_A},\s*{_A},\s*{_A}\)\) = color\({_A},\s*{_A},\s*{_A}\)")
def step_then_pattern_at_eq_color(context, pat_var, px, py, pz, cr, cg, cb):
    pat = getattr(context, pat_var)
    result = pat.pattern_at(Point(parse_math(px), parse_math(py), parse_math(pz)))
    expected = Color(parse_math(cr), parse_math(cg), parse_math(cb))
    assert result == expected, f"{result!r} != {expected!r}"


@then(rf"{_V}\.transform = identity_matrix")
def step_then_pattern_transform_identity(context, var):
    assert getattr(context, var).transform == IDENTITY_MATRIX


@then(rf"{_V}\.transform = translation\({_A},\s*{_A},\s*{_A}\)")
def step_then_pattern_transform_translation(context, var, x, y, z):
    assert getattr(context, var).transform == translation(parse_math(x), parse_math(y), parse_math(z))


@then(rf"c = {_V}")
def step_then_c_eq_var(context, color_var):
    assert context.c == getattr(context, color_var)


@then(rf"c = color\({_A},\s*{_A},\s*{_A}\)")
def step_then_c_eq_color(context, r, g, b):
    expected = Color(parse_math(r), parse_math(g), parse_math(b))
    assert context.c == expected, f"{context.c!r} != {expected!r}"
